File: python/sort/Kth_element.py

a = [1, 4, 7, 3, 11, 9, 2, 6, 4, 7]

def find_element(a, p, q, k):
    element = None

    j = p
    povit = a[q]
    for i in range(p, q+1):
        if a[i] < povit:
            tmp = a[i]
            a[i] = a[j]
            a[j] = tmp

            j += 1

    tmp = a[j]
    a[j] = a[q]
    a[q] = tmp

    print(a)
    print('j = %s, k = %s' % (j, k-1))

    if j == (k-1):
        print('find element')
        print(a)

    if j > (k-1):
        element = find_element(a, p, j-1, k)
    elif j < (k-1):
        element = find_element(a, j+1, q, k)
    else:
        element = a[j]

    if element:
        return element
    

def test():
    tmp = 1
    return tmp
    

def main():
    ret = find_element(a, 0, len(a)-1, 10)
    print(a)
    # ret = test()
    print('value = ', ret)

if __name__ == '__main__':
    main()

File: c/sort/Kth_element.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define N 10
int a[N] = {1, 6, 3, 8, 15, 6, 7, 3, 12, 10};

int find_element(int *a, int p, int q, int k) {

    int element = -1;
    int j = p;
    int povit = a[q];

    for (int i = p; i < q+1; i++) {
        if (a[i] < povit) {
            int tmp = a[i];
            a[i] = a[j];
            a[j] = tmp;
            j++;
        }
    }

    int tmp = a[j];
    a[j] = a[q];
    a[q] = tmp;

    if (j > (k-1))
        element = find_element(a, p, j-1, k);
    else if (j < (k-1))
        element = find_element(a, j+1, q, k);
    else
        element = a[j];
    
    return element;

}

int main() {
    int ret = find_element(a, 0, N-1, 9);
    printf("element is %d\n", ret);
    return 0;
}